### PR TITLE
add a way to change the destination directory of the built files

### PR DESCRIPTION
**Pull request by erwinmombay**
This allows for us to build alternative "builds/experiments" of the source code without overwriting the `dest` folder. This is is useful when deploying these alternative builds/experiments into different directories without having to run the process of deployment multiple times.

### CODE REVIEW

**Comment by erwinmombay** [comment]
@rsimha i'd like to consolidate it as much as possible. theres a few instances where `dist` can't be changed (like embedded in html) but for the most part if i missed something, please point it out.

The hardcoded strings everywhere is a pain 😅

**Comment by rsimha** [comment]
There are ~140 instances of the word `dist` in `build-system/`. A good number of them are instances of a hard-coded directory name and should probably be replaced. Search results [here](https://github.com/ampproject/amphtml/search?q=%22dist%22+in%3Abuild-system&unscoped_q=%22dist%22+in%3Abuild-system).

Maybe this PR can be implemented in two stages, one for compilation code, and a follow up for test code?

File: build-system/compile/compile.js [hunk 1]
```diff
@@ -153,7 +153,7 @@ function compile(entryModuleFilenames, outputDir, outputFilename, options) {
 
     // Add babel plugin to remove unwanted polyfills in esm build
     if (options.esmPassCompilation) {
-      compilationOptions['dest'] = './dist/esm/';
+      compilationOptions['dest'] = `./${OUTPUT_DIR}esm/`;
```
Comment (prateekbh) [inline build-system/compile/compile.js]: Isn't the build pipeline supposed to be windows compatible?
If yes, please use `path.join`

File: build-system/tasks/generate-vendor-jsons.js [hunk 2]
```diff
@@ -50,7 +51,9 @@ function generateVendorJsons() {
 
   return compileJs(srcDir, srcFile, tempPath, compileOptions).then(() => {
     const promises = [];
-    const {ANALYTICS_CONFIG} = require('../../dist/temp-analytics/vendors.js');
+    const {
+      ANALYTICS_CONFIG,
+    } = require(`../../${OUTPUT_DIR}/temp-analytics/vendors.js`);
```
Comment (jonathantyng-amp) [inline build-system/tasks/generate-vendor-jsons.js]: should there be a `/` after `${OUTPUT_DIR}` here?

File: build-system/tasks/visual-diff/index.js [hunk 3]
```diff
@@ -827,7 +828,7 @@ async function performVisualTests() {
 async function ensureOrBuildAmpRuntimeInTestMode_() {
   if (argv.nobuild) {
     const isInTestMode = /AMP_CONFIG=\{(?:.+,)?"test":true\b/.test(
-      fs.readFileSync('dist/v0.js', 'utf8')
+      fs.readFileSync(`${OUTPUT_DIR}/v0.js`, 'utf8')
```
Comment (jonathantyng-amp) [inline build-system/tasks/visual-diff/index.js]: same here

**Comment by jonathantyng-amp** [comment]
There's a path in `compileVendorConfigs()` in `build-system/tasks/vendor-configs.js` that also needs to be updated

File: build-system/sources.js [hunk 4]
```diff
@@ -126,8 +127,16 @@ const CLOSURE_SRC_GLOBS = [
   '!node_modules/core-js/modules/library/**.js',
 ].concat(COMMON_GLOBS);
 
+let destination =
+  typeof argv.destination === 'string' ? argv.destination : './dist/';
+if (!destination.endsWith('/')) {
+  destination = `${destination}/`;
+}
+const OUTPUT_DIR = destination;
```
Comment (zhouyx) [inline build-system/sources.js]: Personally I find it hard to read when `OUTPUT_DIR` comes with the ending `/`. Can we always remove the ending `/` instead?

File: build-system/sources.js [hunk 5]
```diff
@@ -126,8 +127,16 @@ const CLOSURE_SRC_GLOBS = [
   '!node_modules/core-js/modules/library/**.js',
 ].concat(COMMON_GLOBS);
 
+let destination =
+  typeof argv.destination === 'string' ? argv.destination : './dist/';
```
Comment (zhouyx) [inline build-system/sources.js]: So if I understand it correctly the `argv.destination` will be declared in the release script where you will be calling `gulp dist`. I think that's good.
What are we setting the `argv.destination` to?
I think this is good. But are we planning on fix the destination based on experiment channel? Or do we want the experiment to define their own destination folder? Thanks

**Comment by mrjoro** [comment]
Is this PR still active?

**Pull request closed without merging**